File: packages/python/procedurals/target_springs.py
"""Ordered independent target-spring one-step transition independently specified
by the shared motion.target-springs-2d contract.

Motivating note: survey/out/2018/Generativos/araniaaas/notes.md (target force,
velocity and position update). No artistic strength/retention range is
established; see the catalog contract for evidence. Each call is a pure
one-step transition; feed a returned state back in as the next call's state
for animation.
"""
from array import array
import math

__all__ = ["SpringError", "target_springs_2d"]
_MAX_BODIES = 357913941
_MAX_SAFE = 9007199254740991


class SpringError(ValueError):
    """Stable catalog code, with body_index/axis/stage for SPRING_ARITHMETIC_INVALID."""
    def __init__(self, code, body_index=None, axis=None, stage=None):
        super().__init__(code)
        self.code = code
        if code == "SPRING_ARITHMETIC_INVALID":
            self.body_index = body_index
            self.axis = axis
            self.stage = stage


def _fail(code):
    raise SpringError(code)


def _zero(value):
    return 0.0 if value == 0 else value


def _number(value, code):
    if type(value) not in (int, float) or type(value) is bool:
        _fail(code)
    try:
        value = float(value)
    except OverflowError:
        raise SpringError(code) from None
    if not math.isfinite(value):
        _fail(code)
    return _zero(value)


def _strength_of(value):
    n = _number(value, "INVALID_INPUT")
    if n < 0:
        _fail("INVALID_INPUT")
    return n


def _retention_of(value):
    n = _number(value, "INVALID_INPUT")
    if n < 0 or n > 1:
        _fail("INVALID_INPUT")
    return n


def _pair(value, code):
    if type(value) is not list or len(value) != 2:
        _fail(code)
    return value


def _validate_body(supplied):
    if type(supplied) is not dict or set(supplied) != {"position", "velocity", "strength", "retention"}:
        _fail("INVALID_INPUT")
    position = _pair(supplied["position"], "INVALID_INPUT")
    _number(position[0], "INVALID_INPUT")
    _number(position[1], "INVALID_INPUT")
    velocity = _pair(supplied["velocity"], "INVALID_INPUT")
    _number(velocity[0], "INVALID_INPUT")
    _number(velocity[1], "INVALID_INPUT")
    _strength_of(supplied["strength"])
    _retention_of(supplied["retention"])


def _validate_target_pair(supplied):
    target = _pair(supplied, "INVALID_INPUT")
    _number(target[0], "INVALID_INPUT")
    _number(target[1], "INVALID_INPUT")


def _arithmetic(value, body_index, axis, stage):
    if not math.isfinite(value):
        raise SpringError("SPRING_ARITHMETIC_INVALID", body_index, axis, stage)
    return value


class _TargetSprings:
    """Immutable one-step result; see module docstring for the animation pattern."""
    __slots__ = ("_current", "_count")

    def __init__(self, current, count):
        self._current = current
        self._count = count

    @property
    def size(self):
        return self._count

    def _checked_index(self, index):
        if type(index) not in (int, float) or type(index) is bool:
            _fail("INVALID_INDEX")
        if type(index) is float and (not math.isfinite(index) or not index.is_integer()):
            _fail("INVALID_INDEX")
        if index < 0 or index > _MAX_SAFE:
            _fail("INVALID_INDEX")
        index = int(index)
        if index >= self._count:
            raise SpringError("INDEX_OUT_OF_RANGE")
        return index

    @staticmethod
    def _check_output(output, offset):
        if not (type(output) is list or (type(output) is array and output.typecode == "d" and output.itemsize == 8)):
            raise SpringError("INVALID_OUTPUT")
        if type(offset) not in (int, float) or type(offset) is bool:
            raise SpringError("INVALID_OUTPUT")
        if type(offset) is float and (not math.isfinite(offset) or not offset.is_integer()):
            raise SpringError("INVALID_OUTPUT")
        offset = int(offset)
        if offset < 0 or offset > len(output) - 2:
            raise SpringError("INVALID_OUTPUT")
        return offset

    def position_at(self, index):
        """Return a fresh detached [x, y] position."""
        i = self._checked_index(index)
        offset = i * 6
        return [self._current[offset], self._current[offset + 1]]

    def velocity_at(self, index):
        """Return a fresh detached [x, y] velocity."""
        i = self._checked_index(index)
        offset = i * 6
        return [self._current[offset + 2], self._current[offset + 3]]

    def position_into(self, index, output, offset=0):
        """Validate index and destination before writing; return output."""
        i = self._checked_index(index)
        offset = self._check_output(output, offset)
        source = i * 6
        output[offset] = self._current[source]
        output[offset + 1] = self._current[source + 1]
        return output

    def velocity_into(self, index, output, offset=0):
        """Validate index and destination before writing; return output."""
        i = self._checked_index(index)
        offset = self._check_output(output, offset)
        source = i * 6
        output[offset] = self._current[source + 2]
        output[offset + 1] = self._current[source + 3]
        return output

    def strength_at(self, index):
        return self._current[self._checked_index(index) * 6 + 4]

    def retention_at(self, index):
        return self._current[self._checked_index(index) * 6 + 5]

    def to_values(self):
        """Export detached output-schema state; retain no export objects."""
        bodies = []
        offset = 0
        for _ in range(self._count):
            bodies.append({
                "position": [self._current[offset], self._current[offset + 1]],
                "velocity": [self._current[offset + 2], self._current[offset + 3]],
                "strength": self._current[offset + 4],
                "retention": self._current[offset + 5],
            })
            offset += 6
        return {"bodies": bodies}


def _build_state(state):
    if type(state) is not dict or set(state) != {"bodies"}:
        _fail("INVALID_INPUT")
    bodies = state["bodies"]
    if type(bodies) is not list:
        _fail("INVALID_INPUT")
    if len(bodies) > _MAX_BODIES:
        _fail("INVALID_INPUT")
    count = len(bodies)
    for body in bodies:
        _validate_body(body)

    current = array("d", [0.0]) * (count * 6)
    offset = 0
    for body in bodies:
        position = _pair(body["position"], "INVALID_INPUT")
        velocity = _pair(body["velocity"], "INVALID_INPUT")
        current[offset] = _number(position[0], "INVALID_INPUT")
        current[offset + 1] = _number(position[1], "INVALID_INPUT")
        current[offset + 2] = _number(velocity[0], "INVALID_INPUT")
        current[offset + 3] = _number(velocity[1], "INVALID_INPUT")
        current[offset + 4] = _strength_of(body["strength"])
        current[offset + 5] = _retention_of(body["retention"])
        offset += 6
    return current, count


def _build_targets(targets, count):
    if type(targets) is not list or len(targets) != count:
        _fail("INVALID_INPUT")
    for item in targets:
        _validate_target_pair(item)
    captured = array("d", [0.0]) * (count * 2)
    offset = 0
    for item in targets:
        pair = _pair(item, "INVALID_INPUT")
        captured[offset] = _number(pair[0], "INVALID_INPUT")
        captured[offset + 1] = _number(pair[1], "INVALID_INPUT")
        offset += 2
    return captured


def target_springs_2d(config):
    """Advance an ordered independent target-spring state by one explicit logical
    step: target force, position using updated velocity, then velocity retention.

    Implements motion.target-springs-2d 0.1.0 independently of source code.
    """
    if type(config) is not dict or set(config) != {"state", "targets"}:
        _fail("INVALID_INPUT")
    current, count = _build_state(config["state"])
    captured = _build_targets(config["targets"], count)

    next_state = array("d", [0.0]) * (count * 6)
    state_offset = 0
    target_offset = 0
    for body in range(count):
        for axis, axis_offset in (("x", 0), ("y", 1)):
            delta = _arithmetic(captured[target_offset + axis_offset] - current[state_offset + axis_offset], body, axis, "delta")
            force = _arithmetic(delta * current[state_offset + 4], body, axis, "force")
            advanced = _arithmetic(current[state_offset + 2 + axis_offset] + force, body, axis, "advanced")
            position = _arithmetic(current[state_offset + axis_offset] + advanced, body, axis, "position")
            velocity = _arithmetic(advanced * current[state_offset + 5], body, axis, "velocity")
            next_state[state_offset + axis_offset] = _zero(position)
            next_state[state_offset + 2 + axis_offset] = _zero(velocity)
        next_state[state_offset + 4] = current[state_offset + 4]
        next_state[state_offset + 5] = current[state_offset + 5]
        state_offset += 6
        target_offset += 2
    return _TargetSprings(next_state, count)

File: packages/javascript/src/target-springs.js
const MAX_BODIES = 357913941;
const MAX_SAFE = 9007199254740991;

/** Error with one of the stable motion.target-springs-2d contract codes. */
export class SpringError extends Error {
  constructor(code, bodyIndex = null, axis = null, stage = null) {
    super(code);
    this.name = "SpringError";
    this.code = code;
    if (code === "SPRING_ARITHMETIC_INVALID") {
      this.bodyIndex = bodyIndex;
      this.axis = axis;
      this.stage = stage;
    }
  }
}

function fail(code) { throw new SpringError(code); }
function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function zero(value) { return value === 0 ? 0 : value; }
function number(value, code) { if (!finite(value)) fail(code); return zero(value); }
function strengthOf(value) { const n = number(value, "INVALID_INPUT"); if (n < 0) fail("INVALID_INPUT"); return n; }
function retentionOf(value) { const n = number(value, "INVALID_INPUT"); if (n < 0 || n > 1) fail("INVALID_INPUT"); return n; }

function passiveRecord(value, keys, code) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail(code);
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null) fail(code);
  if (Reflect.ownKeys(value).length !== keys.length) fail(code);
  for (const key of keys) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key);
    if (descriptor === undefined || !("value" in descriptor)) fail(code);
  }
  return value;
}
function recordValue(record, key) { return Object.getOwnPropertyDescriptor(record, key).value; }
function passiveArray(value, code) {
  if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype) fail(code);
  return value;
}
function arrayItem(array, index, code) {
  const descriptor = Object.getOwnPropertyDescriptor(array, String(index));
  if (descriptor === undefined || !("value" in descriptor)) fail(code);
  return descriptor.value;
}
function pair(value, code) {
  const array = passiveArray(value, code);
  if (array.length !== 2) fail(code);
  return array;
}
function writableSlot(array, index) {
  const key = String(index);
  const descriptor = Object.getOwnPropertyDescriptor(array, key);
  if (descriptor !== undefined) return "value" in descriptor && descriptor.writable === true;
  if (!Object.isExtensible(array)) return false;
  for (let prototype = Object.getPrototypeOf(array); prototype !== null; prototype = Object.getPrototypeOf(prototype)) {
    if (Object.getOwnPropertyDescriptor(prototype, key) !== undefined) return false;
  }
  return true;
}

function validateBody(supplied) {
  const record = passiveRecord(supplied, ["position", "velocity", "strength", "retention"], "INVALID_INPUT");
  const position = pair(recordValue(record, "position"), "INVALID_INPUT");
  number(arrayItem(position, 0, "INVALID_INPUT"), "INVALID_INPUT");
  number(arrayItem(position, 1, "INVALID_INPUT"), "INVALID_INPUT");
  const velocity = pair(recordValue(record, "velocity"), "INVALID_INPUT");
  number(arrayItem(velocity, 0, "INVALID_INPUT"), "INVALID_INPUT");
  number(arrayItem(velocity, 1, "INVALID_INPUT"), "INVALID_INPUT");
  strengthOf(recordValue(record, "strength"));
  retentionOf(recordValue(record, "retention"));
}
function validateTargetPair(supplied) {
  const target = pair(supplied, "INVALID_INPUT");
  number(arrayItem(target, 0, "INVALID_INPUT"), "INVALID_INPUT");
  number(arrayItem(target, 1, "INVALID_INPUT"), "INVALID_INPUT");
}
function arithmetic(value, bodyIndex, axis, stage) {
  if (!Number.isFinite(value)) throw new SpringError("SPRING_ARITHMETIC_INVALID", bodyIndex, axis, stage);
  return value;
}

class TargetSprings {
  #current; #count;
  constructor(current, count) {
    this.#current = current;
    this.#count = count;
    Object.freeze(this);
  }
  get size() { return this.#count; }

  #checkedIndex(index) {
    if (typeof index !== "number" || !Number.isFinite(index) || index < 0
        || index !== Math.floor(index) || index > MAX_SAFE) fail("INVALID_INDEX");
    if (index >= this.#count) throw new SpringError("INDEX_OUT_OF_RANGE");
    return index;
  }
  #checkOutput(output, offset) {
    if (!Array.isArray(output) && !(output instanceof Float64Array)) throw new SpringError("INVALID_OUTPUT");
    if (!Number.isInteger(offset) || offset < 0 || offset > output.length - 2) throw new SpringError("INVALID_OUTPUT");
    if (!writableSlot(output, offset) || !writableSlot(output, offset + 1)) throw new SpringError("INVALID_OUTPUT");
  }
  positionAt(index) {
    const i = this.#checkedIndex(index), offset = i * 6;
    return [this.#current[offset], this.#current[offset + 1]];
  }
  velocityAt(index) {
    const i = this.#checkedIndex(index), offset = i * 6;
    return [this.#current[offset + 2], this.#current[offset + 3]];
  }
  positionInto(index, output, offset = 0) {
    const i = this.#checkedIndex(index);
    this.#checkOutput(output, offset);
    const source = i * 6;
    output[offset] = this.#current[source];
    output[offset + 1] = this.#current[source + 1];
    return output;
  }
  velocityInto(index, output, offset = 0) {
    const i = this.#checkedIndex(index);
    this.#checkOutput(output, offset);
    const source = i * 6;
    output[offset] = this.#current[source + 2];
    output[offset + 1] = this.#current[source + 3];
    return output;
  }
  strengthAt(index) { return this.#current[this.#checkedIndex(index) * 6 + 4]; }
  retentionAt(index) { return this.#current[this.#checkedIndex(index) * 6 + 5]; }

  toValues() {
    const bodies = [];
    for (let index = 0, offset = 0; index < this.#count; index += 1, offset += 6) {
      bodies.push({
        position: [this.#current[offset], this.#current[offset + 1]],
        velocity: [this.#current[offset + 2], this.#current[offset + 3]],
        strength: this.#current[offset + 4],
        retention: this.#current[offset + 5],
      });
    }
    return { bodies };
  }
}

function buildState(state) {
  const record = passiveRecord(state, ["bodies"], "INVALID_INPUT");
  const suppliedBodies = recordValue(record, "bodies");
  const bodies = passiveArray(suppliedBodies, "INVALID_INPUT");
  if (bodies.length > MAX_BODIES) fail("INVALID_INPUT");
  const count = bodies.length;
  for (let i = 0; i < count; i += 1) validateBody(arrayItem(bodies, i, "INVALID_INPUT"));

  const current = new Array(count * 6).fill(0);
  for (let i = 0, offset = 0; i < count; i += 1, offset += 6) {
    const body = passiveRecord(arrayItem(bodies, i, "INVALID_INPUT"), ["position", "velocity", "strength", "retention"], "INVALID_INPUT");
    const position = pair(recordValue(body, "position"), "INVALID_INPUT");
    const velocity = pair(recordValue(body, "velocity"), "INVALID_INPUT");
    current[offset] = number(arrayItem(position, 0, "INVALID_INPUT"), "INVALID_INPUT");
    current[offset + 1] = number(arrayItem(position, 1, "INVALID_INPUT"), "INVALID_INPUT");
    current[offset + 2] = number(arrayItem(velocity, 0, "INVALID_INPUT"), "INVALID_INPUT");
    current[offset + 3] = number(arrayItem(velocity, 1, "INVALID_INPUT"), "INVALID_INPUT");
    current[offset + 4] = strengthOf(recordValue(body, "strength"));
    current[offset + 5] = retentionOf(recordValue(body, "retention"));
  }
  return { current, count };
}

function buildTargets(targets, count) {
  const pairs = passiveArray(targets, "INVALID_INPUT");
  if (pairs.length !== count) fail("INVALID_INPUT");
  for (let i = 0; i < pairs.length; i += 1) validateTargetPair(arrayItem(pairs, i, "INVALID_INPUT"));
  const captured = new Array(count * 2).fill(0);
  for (let i = 0, offset = 0; i < pairs.length; i += 1, offset += 2) {
    const p = pair(arrayItem(pairs, i, "INVALID_INPUT"), "INVALID_INPUT");
    captured[offset] = number(arrayItem(p, 0, "INVALID_INPUT"), "INVALID_INPUT");
    captured[offset + 1] = number(arrayItem(p, 1, "INVALID_INPUT"), "INVALID_INPUT");
  }
  return captured;
}

/**
 * Advance an ordered independent target-spring state by one explicit logical step:
 * target force, position using updated velocity, then velocity retention.
 * Implements motion.target-springs-2d 0.1.0 independently of source code. Motivating
 * sketch: survey/out/2018/Generativos/araniaaas; no artistic strength/retention range
 * is established, see the catalog contract for evidence. Each call is a pure one-step
 * transition; feed a returned state back in as the next call's state for animation.
 */
export function targetSprings2D(input) {
  const record = passiveRecord(input, ["state", "targets"], "INVALID_INPUT");
  const { current, count } = buildState(recordValue(record, "state"));
  const captured = buildTargets(recordValue(record, "targets"), count);

  const next = new Array(count * 6).fill(0);
  for (let body = 0, stateOffset = 0, targetOffset = 0; body < count; body += 1, stateOffset += 6, targetOffset += 2) {
    for (const [axis, axisOffset] of [["x", 0], ["y", 1]]) {
      const delta = arithmetic(captured[targetOffset + axisOffset] - current[stateOffset + axisOffset], body, axis, "delta");
      const force = arithmetic(delta * current[stateOffset + 4], body, axis, "force");
      const advanced = arithmetic(current[stateOffset + 2 + axisOffset] + force, body, axis, "advanced");
      const position = arithmetic(current[stateOffset + axisOffset] + advanced, body, axis, "position");
      const velocity = arithmetic(advanced * current[stateOffset + 5], body, axis, "velocity");
      next[stateOffset + axisOffset] = zero(position);
      next[stateOffset + 2 + axisOffset] = zero(velocity);
    }
    next[stateOffset + 4] = current[stateOffset + 4];
    next[stateOffset + 5] = current[stateOffset + 5];
  }
  return new TargetSprings(next, count);
}
